File: app/src/main/java/com/example/eldho/workmanagersample/WorkerClass.java
package com.example.eldho.workmanagersample;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Data;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

/**
 * We can pass multiple work request to work manager, if we pass several work req to a single method call, work manager do it in parallel
 */

//Worker : The work needed to be done is defined here
public class WorkerClass extends Worker {
    private Context mContext;
    private static final String TAG = "WorkerClass";
    public static final String TASK_DESC = "task_desc";
    public static final String IS_TO_BE_DONE = " isToBeDone";
    public static final String TASK_OUTPUT = "taskOutPut";
    public static final String IS_DONE = "isDone";

    public WorkerClass(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
        mContext = context;
    }

    @NonNull
    @Override
    /**Inside this func we do our work that we need to exec in background*/
    public Result doWork() {
        Data data = getInputData(); // get the data which are passing to workRequest
        String desc = data.getString(TASK_DESC);
        try {
            // Work here is to display a notification
            NotificationUtils.createNotifications(mContext, "Success", desc);
            Data result = getResultForSending(true, "Success"); // Sending success data/Result back to activity/fragment
            return Result.success(result); // if work is success
        } catch (Throwable throwable) {

            // Technically WorkManager will return Result.failure()
            // but it's best to be explicit about it.
            // Thus if there were errors, we're return FAILURE
            Log.e(TAG, "Error ", throwable);
            Data result = getResultForSending(false, "Failure");
            return Result.failure(result);
//          return Result.retry(); // if we want to retry the task
        }
    }

    // Sending success data/Result back to activity/fragment which then observed in `getWorkInfoByIdLiveData`
    private Data getResultForSending(boolean isDone, String taskOutput) {
        Data data = new Data.Builder()
                .putString(WorkerClass.TASK_OUTPUT, taskOutput) //Note : We can send almost all data types
                .putBoolean(WorkerClass.IS_DONE, isDone)
                .build();
        return data;
    }
}

File: app/src/main/java/com/example/eldho/workmanagersample/workers/WorkerClass.kt
package com.example.eldho.workmanagersample.workers

import android.content.Context
import android.util.Log
import androidx.hilt.work.HiltWorker
import androidx.work.CoroutineWorker
import androidx.work.Data
import androidx.work.WorkerParameters
import androidx.work.workDataOf
import com.example.eldho.workmanagersample.MainActivity
import com.example.eldho.workmanagersample.notification.NotificationUtils
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext


@HiltWorker // Dependency injection in worker class is not a straight forward thing we needed to use hilt work for injecting dependencies into worker class
class WorkerClass @AssistedInject constructor(
    @Assisted private val appContext: Context,
    @Assisted workerParams: WorkerParameters,
    private val notificationUtils: NotificationUtils
) :
    CoroutineWorker(appContext, workerParams) {

    /**
     * doWork() method is run async (Because of CoroutineWorker, else if we use Worker it will run synchronously ) on a background thread provided by WorkManager
     *  */
    override suspend fun doWork(): Result = withContext(Dispatchers.Default) {
        val data = inputData // get the data which are passing to workRequest
        val desc = data.getString(MainActivity.TASK_DESC)
        return@withContext try {
            // Work here is to display a notification
            notificationUtils.createNotifications("Success", desc)

            val result = getSendResult(
                true,
                "Success") // Sending success data/Result back to activity/fragment
            Result.success(result) // if work is success
        } catch (throwable: Throwable) {

            // Technically WorkManager will return Result.failure()
            // but it's best to be explicit about it.
            // Thus if there were errors, we're return FAILURE
            Log.e(TAG, "Error ", throwable)
            val result = getSendResult(false, "Failure")
            Result.failure(result)
            //Result.retry()  // if we want to retry the task
        }
    }

    // Sending success data/Result back to activity/fragment which then observed in `getWorkInfoByIdLiveData`
    private fun getSendResult(isDone: Boolean, taskOutput: String): Data {
        return workDataOf(
            TASK_OUTPUT to taskOutput,
            IS_DONE to isDone
        )
    }

    companion object {
        private const val TAG = "WorkerClass"
        const val TASK_OUTPUT = "taskOutPut"
        const val IS_DONE = "isDone"
    }
}
